util: improve FindByte() performance

LarryRuane opened the pull request:
This PR is strictly a performance improvement; there is no functional change. The `CBufferedFile::FindByte()` method searches for the next occurrence of the given byte in the file. Currently, this is done by explicitly inspecting each byte in turn. This PR takes advantage of `std::find()` to do the same more efficiently, improving its CPU runtime by a factor of about 25 in typical use.

LarryRuane commented:
This PR was suggested by @hebasto https://github.com/bitcoin/bitcoin/pull/16981#issuecomment-669658597 (thank you!)

promag commented on src/streams.h:
```diff
@@ -832,12 +832,21 @@ class CBufferedFile
-            nReadPos++;
+            size_t n = vchBuf.size() - start;
+            if (n > nSrcPos - nReadPos)
+                n = nSrcPos - nReadPos;
```
nit, join with above line.

promag reviewed (commented):
The performance gain looks substantial - didn't verified.

hebasto commented:
@LarryRuane 
> ... improving its CPU runtime by a factor of about 25 in typical use.

How was it measured?

laanwj commented:
Any idea *why* this is so much faster? As far as I know, there is no faster algorithm to look for the first occurrence of a single byte in a memory array than a linear iteration over it. I'd expect `std::find` of a byte to simply unroll into a loop.

LarryRuane commented:
> How was it measured?

I ran:
```
time src/test/test_bitcoin --run_test=streams_tests/streams_findbyte
```
with master (1m20s) and with master + PR (3s).

> Any idea _why_ this is so much faster? 

I think you're correct that there's no faster way than a loop with runtime proportional to the number of bytes to scan, but I assume `std::find()` on a char vector is highly optimized, probably using `memchr()` or `memcmp()`, which are implemented in assembly language. Also, the master version does a few things each byte (testing `nReadPos == nSrcPos`, remainder calculation (`%`), incrementing `nReadPos`) that the PR does once for each large run of bytes.

I just noticed that the repetition count on the test is set to a large number (50000000) and I meant to reduce it for the commit (3 seconds is too long to add to the unit test suite). I'll reduce that number in force push in a minute. This test doesn't really need to be in this PR (`FindByte()`'s functionality is tested very well in another test), but it helps reviewers verify the performance improvement.

LarryRuane commented:
Force-push ([diff](https://github.com/bitcoin/bitcoin/compare/ab412ec3071dd056bad362aa78c88d02d5a46c3b..a31aa3255499ba8878ba98e4ac5a22a408c68da0)) small fix to the test, so it doesn't take 3 seconds to run.

laanwj commented:
> but I assume std::find() on a char vector is highly optimized, probably using memchr() or memcmp(), which are implemented in assembly language

That's true, it's possible to optimize that with assembly language (definitely with specific instruction sets).

it still surprises me because you'd expect the I/O, to read the data from disk, to dominate greatly in the block importing. Not looking for a character already in memory! It just seems out of proportion.

glozow committed:
[bench] add streams findbyte

gzhao408 commented on src/test/streams_tests.cpp:
```diff
@@ -445,4 +445,29 @@ BOOST_AUTO_TEST_CASE(streams_buffered_file_rand)
+    fwrite(&b, 1, 1, file);
+    rewind(file);
+    CBufferedFile bf(file, fileSize * 2, fileSize, 0, 0);
+    for (int rep = 0; rep < 100; ++rep) {
```
I don't fully understand how the performance increase is so significant, but why not a bench if you're worried about burdening the unit tests? I [tried](https://github.com/gzhao408/bitcoin/blob/c3f7075f538cca89f44923d9489403581650edc8/src/bench/streams_findbyte.cpp) to do this but I must be doing something wrong because I can't seem to reproduce the speedup. 😞

sipa replied:
The performance impact may be very compiler/architecture/stdlib dependent. I'm kind of surprised std::find has optimizations beyond the naive loop implementation in the first place on some platforms, so I certainly wouldn't be surprised if others don't have it.

LarryRuane replied:
@gzhao408, thank you, I wasn't aware of `bench`. I suspect the iteration count, 100, is far too low and the difference is swamped by the noise. I increased the iteration count to 10m (`1e7`) and it showed the expected difference, master: 1,659.30 ns/op, PR: 52.66 ns/op (ratio is about 31).

I just force-pushed ([diff](https://github.com/bitcoin/bitcoin/compare/a31aa3255499ba8878ba98e4ac5a22a408c68da0..8b07e17cb2a20db8c803cfbb8491ae51a75bd450)) to remove the unit test (which was only for benchmarking, not really testing anything) and cherry-pick Gloria's benchmark. I added one more commit to increase the iteration count.

LarryRuane commented:
> it still surprises me because you'd expect the I/O, to read the data from disk, to dominate greatly in the block importing.

`FindByte()` only reads from disk by calling `Fill()` (when the buffer is empty), which is rare. In this test, `Fill()` gets called only once, the first time `FindByte()` runs, because I wanted to isolate the modified part of the code.

LarryRuane committed:
util: improve streams.h:FindByte() performance

Avoid use of the expensive mod operator (%) when calculating the
buffer offset. No functional difference.

Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

sipa commented:
@gmaxwell pointed out to me why this is so much faster: it's not that std::find is amazing, but that the original code (which I wrote in 2012, it seems!) is doing a modulus operation for every character (which is often orders of magnitude slower than the byte comparison or addition/subtraction).

Thinking about this a bit more high level: the end goal is just to scan quickly for the 4-byte network magic in a file. If this is relevant for performance (and it seems it may be), it may be better to have a function that does exactly that in CBufferedFile, rather than a search for one byte + memcmp. `std::search` is probably what you want.

LarryRuane commented:
Here's a version that's very close in implementation to master that eliminates the modulus operation:
```
    void FindByte(char ch) {
        size_t start = nReadPos % vchBuf.size();
        while (true) {
            if (nReadPos == nSrcPos)
                Fill();
            if (vchBuf[start] == ch)
                break;
            nReadPos++;
            start++;
            if (start >= vchBuf.size()) start = 0;
        }
```
and that does make a significant difference; `bench_bitcoin` reports 314 ns/op for this version. (My laptop isn't set up for very accurate CPU benchmarking, but my results are pretty consistent, varying by less than 1% across multiple runs.) That's about 5.3 times as fast as master (1,659.30 ns/op, as [mentioned above](https://github.com/bitcoin/bitcoin/pull/19690#discussion_r468201206)). The PR (52.66 ns/op) is still about 6 times faster than this version. So this version is right about in the middle (in terms of ratios) between master and this PR.

I do like @sipa's suggestion to generalize `FindByte()` to search for a given sequence of bytes (maybe called `FindBytes()` or `Search()`, suggestions welcome); that's a much nicer interface. I'll push a commit to do that within the next few hours.

LarryRuane commented:
I just added a new commit (25413ab, can squash later) to implement @sipa's suggestion to add a method to `CBufferedFile` to find a sequence of bytes, rather than just one byte, as `FindByte()` does. This simplifies `LoadExternalBlockFile()`; the overall code base isn't simpler, but it encapsulates complexity nicely within the `CBufferedFile` class. 

It didn't work out to use `std::search()` because that function requires the bytes being searched to be consecutive in memory -- that works if the entire file is read into contiguous memory. But `CBufferedFile` implements a circular buffer. The bytes we're searching for might be partially in and partially out of memory (bytes are read on demand), and even if all are in memory, they may be split between the end of the buffer and the beginning (because of wraparound), so trying to use `std::search` gets complicated. `FindByte()` doesn't have these problems since it's looking for a single byte.

adamjonas commented:
Looks like one of the sanitizers finds an integer conversion/truncation problem in the changed code:

```

LarryRuane commented:
Thanks, @adamjonas, force-pushed ([diff](https://github.com/bitcoin/bitcoin/compare/25413abf6826015893538f285f837ad55e86ef6d..8f2e8c27bb9bb3ebb9f094129ce4a6019c46cfc0)) a fix for that signed-unsigned CI failure, added another unit test, some improvements to `Search()`, some `clang-format-diff.py` cleanups.

elichai commented on src/bench/streams_findbyte.cpp:
```diff
@@ -0,0 +1,33 @@
+static void FindByte(benchmark::Bench& bench)
+{
+    // Setup
+    FILE* file = fsbridge::fopen("streams_tmp", "w+b");
```
Could we maybe use something like `memfd_create(2)` for the benchmark, to decrease I/O noise?
Anyone knows if there is a windows equivalent or a higher abstraction in boost::filesystem?

elichai replied:
`fmemopen(3)` also looks interesting

elichai replied:
Maybe this can give us that? https://www.boost.org/doc/libs/1_72_0/libs/iostreams/doc/classes/mapped_file.html

LarryRuane replied:
Thanks for the suggestions, but I don't think this matters, because the file IO (read) occurs only on the very first benchmark loop iteration; after that, the data is in memory and there is no IO at all. Each iteration repositions the stream pointer to zero (`bf.SetPos(0)`) and then searches forward for the value 1 (`bf.FindByte(1)`), which is 200 bytes away. But after the first iteration, all of these bytes are in memory, and we're just moving the position between 0 and 200. 

The reason for the 200, by the way, is that with random data, searching for a random byte value the average distance would be half of 256. But the data in `blk.dat` files (which is the only use of this code currently) isn't quite random; zero and `0xff` occur more often (and we never search for those values). So I guessed that 200 is close to a typical distance that this function would move through before finding the requested byte. Maybe I should explain these points in a comment in `bench/streams_findbyte.cpp`.

laanwj commented:
Code review ACK 8f2e8c27bb9bb3ebb9f094129ce4a6019c46cfc0, this looks like a better abstraction, and it's an improvement not to do a modulus for every byte.

LarryRuane commented:
Force-pushed to rationalize the commits, so I think it's in a good state for merging. I reordered the commit that adds the benchmark test, b576994bea620c7d065dd558c93222391151c9ff, to be first, so it's easy for reviewers to checkout that commit and run a pre-PR benchmark test:

sipa commented:
Code review ACK 566c2e2eec1e84af5143f1554613a89192e29433. I think it'd be good to move the refactoring to FindByte in the last commit to a separate commit.

LarryRuane commented:
Force-pushed to implement latest [review suggestion](https://github.com/bitcoin/bitcoin/pull/19690#issuecomment-739059080), changes are more cleanly separated among the commits (no code changes overall), thanks @sipa.

hebasto commented on src/bench/streams_findbyte.cpp:
```diff
@@ -0,0 +1,33 @@
+    fs::remove("streams_tmp");
+}
+
+BENCHMARK(FindByte);
```
b576994bea620c7d065dd558c93222391151c9ff, nit: Add EOL
![Screenshot from 2020-12-05 16-27-26](https://user-images.githubusercontent.com/32963518/101245636-d490ce00-3716-11eb-87e3-090b962e850f.png)

hebasto commented on src/test/streams_tests.cpp:
```diff
@@ -329,12 +329,80 @@ BOOST_AUTO_TEST_CASE(streams_buffered_file)
+    BOOST_CHECK_EQUAL(bf.GetPos(), 31U);
+
+    // If the byte isn't found, end of file.
+    try {
+        bf.FindByte(99);
+        BOOST_CHECK(false);
+    } catch (const std::exception& e) {
+        BOOST_CHECK(strstr(e.what(),
+                           "CBufferedFile::Fill: end of file") != nullptr);
+    }
```
134de90f2b64ede6dbd812a9a88b00708685394d
```suggestion
    BOOST_CHECK_EXCEPTION(bf.FindByte(99), std::ios_base::failure, HasReason("CBufferedFile::Fill: end of file"));```

hebasto commented on src/test/streams_tests.cpp:
```diff
@@ -329,12 +329,80 @@ BOOST_AUTO_TEST_CASE(streams_buffered_file)
+    // The bytes must be consecutive, this won't be found.
+    {
+        unsigned char needle[2] = {36, 38};
+        try {
+            bf.Search(needle, 2);
+            BOOST_CHECK(false);
+        } catch (const std::exception& e) {
+            BOOST_CHECK(strstr(e.what(),
+                               "CBufferedFile::Fill: end of file") != nullptr);
+        }
```
134de90f2b64ede6dbd812a9a88b00708685394d
```suggestion
        BOOST_CHECK_EXCEPTION(bf.Search(needle, 2), std::ios_base::failure, HasReason("CBufferedFile::Fill: end of file"));
```

hebasto reviewed (changes requested):
Concept ACK.

hebasto commented on src/streams.h:
```diff
@@ -718,7 +718,7 @@ class CBufferedFile
-    std::vector<char> vchBuf; //!< the buffer
+    std::vector<unsigned char> vchBuf; //!< the buffer
 
 protected:
```
While the `CBufferedFile` class is already touched, maybe make `CBufferedFile::Fill` private by removing this line?

practicalswift commented:
@LarryRuane 

Thanks for providing promising micro benchmarks. AFAICT `FindByte` is only used when doing `-reindex` and `-loadblock`. Do you have any macro numbers too where the gains from switching to this improved version are measured in terms of the overall speed-up of said operations?

Another thing: when fuzzing `CBufferedFile` I noticed that calling `FindByte` after a failed `SetPos` call may result in an infinite loop. Does the new code have that gotcha too? If not we can drop the `setpos_fail` logic in `src/test/fuzz/buffered_file.cpp` :)

hebasto commented:
> Another thing: when fuzzing `CBufferedFile` I noticed that calling `FindByte` after a failed `SetPos` call may result in an infinite loop. Does the new code have that gotcha too? If not we can drop the `setpos_fail` logic in `src/test/fuzz/buffered_file.cpp` :)

Also some failed tests:
- https://cirrus-ci.com/task/4540479469518848
- https://bitcoinbuilds.org/?job=34de38c2-b334-4dde-b067-462b3e62200b

LarryRuane commented:
@practicalswift
> Do you have any macro numbers too where the gains from switching to this improved version are measured in terms of the overall speed-up of [`-reindex` and `-loadblock`]?

I don't think there will be any measurable improvement in the sunny-day case, because the byte being searched for is the very first byte at the stream pointer. So no looping is needed (and this PR only improves the performance of that loop). But if there is corruption, damage in the `blk.dat` files, then `FindByte()` is being used to "resynchronize" with the stream, that is, to locate the start of the next valid block. I don't think that happens often in practice. But another advantage of this PR is that `FindByte()` may be used for something else in the future which does require actual looping in the sunny-day case.

> Another thing: when fuzzing `CBufferedFile` I noticed that calling `FindByte` after a failed `SetPos` call may result in an infinite loop. Does the new code have that gotcha too?

TL;DR: No. We can remove that check.

Details: I wanted to understand this better, so I checked out master immediately after your PR #19143 was merged, the commit where you added the `SetPos()` fuzzing (and the `setpos_fail` flag) (commit 614e0807a8137d82832aea45e4864b424f71f698), and did some investigation, and discovered that a failed `SetPos()` isn't what caused the infinite loop. The reason the `setpos_fail` flag seemed to fix the problem is that `SetPos()` typically fails very early in the fuzzing run, and that disables further `FindByte()` calls, after which the problem can't occur. (I didn't see anything that would _prevent_ the infinite loop, but it's just unlikely.)

The actual source of the problem is in the [`Seek()` call](https://github.com/bitcoin/bitcoin/blob/614e0807a8137d82832aea45e4864b424f71f698/src/test/fuzz/buffered_file.cpp#L44). The [implementation](https://github.com/bitcoin/bitcoin/blob/614e0807a8137d82832aea45e4864b424f71f698/src/streams.h#L817) of that method calls the standard library `fseek()`, then, if that succeeds, `ftell()`, which is supposed to return the byte offset into the file of the `FILE` pointer. It sometimes returns `-1` (which indicates error). That seems wrong, because if the `fseek()` is successful, then `ftell()` should return whatever offset we just seeked to. (I'm not sure why `Seek()` even called `ftell()`, since it already knows that offset.) This `-1` (error) return is stored into both the `CBufferedFile`'s read and write offsets (`nReadPos` and `nSrcPos`), which are unsigned, so things are messed up from here. (I was able to see why this caused an infinite loop, but it's not worth explaining here.)

Anyway, this is all irrelevant now, because in #19593, @hebasto removed `CBufferedFile::Seek` because it wasn't being used. So, yes, we should remove this check (I did remove it and verified that there is no infinite loop), and I'll do that in this PR.

Another thing I'd like to do is add fuzz testing for the new method `CBufferedFile::Search()` (it can be added to the [existing test](https://github.com/bitcoin/bitcoin/blob/cef2efafcecc4d3d60421955bbd4c4fcebbcb86d/src/test/fuzz/buffered_file.cpp#L18)). That's why I changed this PR status to DRAFT. Also, I'll look into if those failed tests are real failures. Thanks for all your help!

LarryRuane commented:
Force-pushed the latest review suggestions:
- remove unneeded `protected:` tag
- remove unneeded `setpos_fail` from fuzz test
- add fuzz test for new `CBufferedFile::Search()` method
- simplify checking for expected exception in unit test (use `BOOST_CHECK_EXCEPTION`)

@hebasto 
> Also some failed tests

I could not figure out those failures; I don't see how they have any relation to this PR.

hebasto reviewed (changes requested):
Approach ACK 918de11daa4be9995122c5e4b0284285b27e0e8f

While testing this branch on top of the master still having a [fuzzing problem](https://cirrus-ci.com/build/5192315515961344): `load_external_block_file` timeout. An infinite loop?

If other reviewers are not opposed, I'd suggest to rebase this PR on top of the recent CI changes, in order to make this problem obvious to all reviewers.

As `bench/streams_findbyte.cpp` is a new code, I believe it should follow our Developer Notes (header including and naming). If it'd be re-touched the adding an EOL could be moved from the third commit into the first one. Also the first commit message could contain a verb "Add" :)

To get this PR closer to merging, maybe just drop changes that introduce the `CBufferedFile::Search` member function, as the performance issue has been resolved in the second commit?

LarryRuane commented:
@hebasto, thank you for your careful review! I force-pushed your suggestions except for moving `nReadPos % vchBuf.size();` inside the loop. That does reduce the amount of code slightly, but the benchmark test showed that the time per operation consistently increased from 55ns to 67ns. That's still very good (much better than without this PR), but as long as we're changing this code, we may as well make it as fast as possible (within reason). 

I added you as a co-author, hope that's okay. I found better names for `i` and `n`, thanks.

I took your suggestion to remove the new method, `CBufferedFile::Search()` and leave that for a later PR. That was a good suggestion because now this PR is small and more focused.

It will be interesting to see if now the fuzzing passes. I don't understand how you determined (from the [link](https://cirrus-ci.com/build/5192315515961344) you gave) that the problem is in `load_external_block_file`. When I run that locally (with no arguments), it succeeds normally. Is there a way I can run exactly the same test to debug the problem? I see it uses some fuzzing assets; could I download those?

hebasto commented:
> It will be interesting to see if now the fuzzing passes. I don't understand how you determined (from the [link](https://cirrus-ci.com/build/5192315515961344) you gave) that the problem is in `load_external_block_file`.

Job is timed out and the `load_external_block_file` is not listed in the log.

hebasto reviewed (approved):
ACK 2f772934fe4eb7a10667a89afc5884b3f61a6ab0

Results of `src/bench/bench_bitcoin -filter=FindByte` on my machine:
- master (38176dc66523881d45705fe8f96b9da76ec9981b) + this PR
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               64.12 |       15,596,723.38 |    0.4% |      0.00 | `FindByte
```
- master (38176dc66523881d45705fe8f96b9da76ec9981b) + this PR w/o the top commit
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            1,188.64 |          841,300.44 |    0.1% |      0.00 | `FindByte`
```

CI tests in my personal account [passed](https://cirrus-ci.com/build/6401312118013952).

sipa commented:
Rebase to get new CI?

LarryRuane commented:
Force-pushed rebase, and also simplified the code significantly. Here's the previous (more complex) version of this PR, just for the record:

<details>
<summary>Click to expand patch</summary>

```
--- a/src/streams.h
+++ b/src/streams.h
@@ -795,12 +795,17 @@ public:
 
     //! search for a given byte in the stream, and remain positioned on it
     void FindByte(char ch) {
+        size_t buf_offset = nReadPos % vchBuf.size();
         while (true) {
-            if (nReadPos == nSrcPos)
-                Fill();
-            if (vchBuf[nReadPos % vchBuf.size()] == ch)
-                break;
-            nReadPos++;
+            if (nReadPos == nSrcPos) Fill();
+            const size_t len = std::min<size_t>(vchBuf.size() - buf_offset, nSrcPos - nReadPos);
+            const auto it_start = vchBuf.begin() + buf_offset;
+            const auto it_find = std::find(it_start, it_start + len, ch);
+            const size_t inc = it_find - it_start;
+            nReadPos += inc;
+            if (inc < len) break;
+            buf_offset += inc;
+            if (buf_offset >= vchBuf.size()) buf_offset = 0;
         }
     }
 };
```
</details>

The previous version is _slightly_ faster than the current patch, but it's not worth the added complexity. I re-ran the benchmark (added by the first commit in this PR), and the nanoseconds per operation (on my middle-range laptop) is 1,650 on `master` and 315 with this PR.

stickies-v commented on src/bench/streams_findbyte.cpp:
```diff
@@ -0,0 +1,34 @@
+    b = 1;
+    fwrite(&b, 1, 1, file);
+    rewind(file);
+    CBufferedFile bf(file, file_size * 2, file_size, 0, 0);
```
It's super small either way, but I don't think the buffer needs to be twice the file size?
```suggestion
    CBufferedFile bf(file, file_size + 1, file_size, 0, 0);
```

stickies-v commented on src/bench/streams_findbyte.cpp:
```diff
@@ -0,0 +1,34 @@
+    // Setup
+    FILE* file = fsbridge::fopen("streams_tmp", "w+b");
+    const size_t file_size = 200;
+    uint8_t b = 0;
+    for (size_t i = 0; i < file_size; ++i) {
+        fwrite(&b, 1, 1, file);
+    }
+    b = 1;
+    fwrite(&b, 1, 1, file);
```
Slightly more concise and readable imo:
```suggestion
    uint8_t data[file_size] = {0};
    data[file_size-1] = 1;
    fwrite(&data, sizeof(uint8_t), file_size, file);
```

stickies-v replied:
Less memory efficient, but at a file size of 200 bytes I think optimizing for readability is worthwhile?

stickies-v reviewed (commented):
Approach ACK 0238e5e548d221299ccf57bb8e9633507320997f

On my older x86_64 MacOS laptop, I see very significant performance improvements, speeding up the benchmark from ~1900 ns/op to ~410 ns/op. On my newer M1 Pro MacOS laptop, the difference is much less pronounced: ~940 ns/op -> ~850 ns/op. Did plenty of iterations on both machines, results seem stable.

Given the simplicity and limited scope of this PR, I think this is still a worthwhile improvement (at no extra maintenance cost) even if it only benefits certain platforms.

Left a few comments, I think the bench # iterations needs to be reduced before merging, everything else is optional (but seems straightforward to implement too)

achow101 commented:
Are you still working on this?

There is a silent merge conflict with master, and also unresolved comments.

LarryRuane commented:
> Are you still working on this?

Yes, I think this is still worth doing, I'll refresh today, thanks for the ping.

LarryRuane commented:
Force pushed to resolve a hidden conflict, also addressed all review comments.

stickies-v commented:
Curiously, this PR is now a significant performance degradation (152ns/op -> 417 ns/op) on my M1 Pro. I repeated the benchmarks on the same commits as my last review, and am getting the same performance degradation there now, too. According to my logs I was on macOS 12.4 on both dates. Results are stable enough between runs.

On my x86 (Debian Bullseye) machine, I'm still seeing significant performance improvement (1.717 ns/op -> 338 ns/op), as before.

<details>
<summary>bench results</summary>

M1 Pro @ 5842d92c8 (unpatched)
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              151.60 |        6,596,094.28 |    0.4% |      0.01 | `FindByte`
```

M1 Pro @ 08cf6a9ed812 (patched)
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              417.00 |        2,398,083.95 |    0.2% |      0.01 | `FindByte`
```

x86 (Debian Bullseye) @ 5842d92c8 (unpatched)
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            1,716.73 |          582,504.42 |    0.0% |      0.01 | `FindByte`
```

x86 (Debian Bullseye) @ 08cf6a9ed812 (patched)
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              337.99 |        2,958,650.44 |    1.5% |      0.01 | `FindByte`
```
</details>

LarryRuane commented:
> Curiously, this PR is now a significant performance degradation

As part of the last force-push I made this change:
```diff
--- b/src/bench/streams_findbyte.cpp
+++ a/src/bench/streams_findbyte.cpp
@@ -18,7 +18,7 @@ static void FindByte(benchmark::Bench& bench)
     rewind(file);
     CBufferedFile bf(file, /*nBufSize=*/file_size + 1, /*nRewindIn=*/file_size, 0, 0);
 
-    bench.minEpochIterations(1e7).run([&] {
+    bench.run([&] {
         bf.SetPos(0);
         bf.FindByte(1);
     });
```
Is it possible the test isn't running enough iterations to get an accurate measurement? Although, still, you wouldn't expect the patch to be _worse_!

Can you try with `minEpochIterations(1e5)` (or even back to `1e7`) and see if you get results consistent with before? I probably should restore the iterations count (and set it to 1e5 as you suggest).

I get consistently 136 ns/op with the patch, and 300 ns/op without the patch, with either iterations set to `1e5` or `1e7`. This is Debian Ubuntu jammy AMD x86 desktop.

stickies-v commented:
> Can you try with `minEpochIterations(1e5)` (or even back to `1e7`) and see if you get results consistent with before?

I'd tried that already, but the results are pretty much the same.

john-moffett reviewed (commented):
I think it's mainly an architecture issue. I'm seeing similar performance degradation on my M1 to what @stickies-v reported. It's possible there could be pipeline issues with the conditional selection instructions that get generated by this line:

`if (++buf_offset >= vchBuf.size()) buf_offset = 0;`

The `std::find` version (eg - https://github.com/bitcoin/bitcoin/commit/2f772934fe4eb7a10667a89afc5884b3f61a6ab0) does not have the same issue, and I see significant performance gains in both x86 and ARM. Specifically, on my M1, master is 145 ns/op, 08cf6a9ed812c122e9d47cccee09e03d8139764c is 420 ns/op, and the `std::find` version is 81 ns/op.

My suggestion would be to switch back to that version.

achow101 commented:
ACK dacd3316ca0279c56d12372957633b73a999b5b2

stickies-v reviewed (commented):
Approach ACK dacd3316c

I'm now seeing bench performance improvement on my M1 again, from ~150ns/op -> ~85ns/op.

LarryRuane commented:
Force-pushed for review comments (thanks, @stickies-v), verified benchmark performance is unchanged (ns/op with PR: 34.76, without PR: 302). Summary of force-push changes:
- change `FindByte()` argument type from `uint8_t` to `std::byte`)
- add "exception" comment before call to `Fill()`
- add comment suggesting to avoid mod (%) operator within loop
- changed assignment statements to use more modern braces syntax

stickies-v reviewed (approved):
ACK 0fe832c4a

achow101 commented:
ACK 0fe832c4a4b2049fdf967bca375468d5ac285563

john-moffett reviewed (approved):
ACK 0fe832c4a4b2049fdf967bca375468d5ac285563

achow101 commented:
Silent merge conflict with master:

```
../../../src/bench/streams_findbyte.cpp:7:10: fatal error: fs.h: No such file or directory
    7 | #include <fs.h>
      |          ^~~~~~
compilation terminated.
```

LarryRuane commented:
Force pushed rebase to fix hidden merge conflict, thanks @achow101

stickies-v reviewed (approved):
re-ACK 72efc26439da9a1344a19569fb0cab01f82ae7d1

Verified that the only difference is to include `<util/fs.h>` instead of `<fs.h>` (introduced by https://github.com/bitcoin/bitcoin/pull/27254/commits/00e9b97f37e0bdf4c647236838c10b68b7ad5be3)

```range-diff
% git range-diff HEAD~2 0fe832c4a4b2049fdf967bca375468d5ac285563 HEAD
1:  5842d92c8 ! 1:  604df63f6 [bench] add streams findbyte
    @@ src/bench/streams_findbyte.cpp (new)
     +
     +#include <bench/bench.h>
     +
    -+#include <fs.h>
    ++#include <util/fs.h>
     +#include <streams.h>
     +
     +static void FindByte(benchmark::Bench& bench)
2:  0fe832c4a = 2:  72efc2643 util: improve streams.h:FindByte() performance
```

achow101 commented:
re-ACK 72efc26439da9a1344a19569fb0cab01f82ae7d1
